File: sample/sort/shell-sort.hs

gen :: Int -> [Int]
gen x
  | x == 0 = [0]
  | otherwise = x: gen (x `div` 2)

shellSort :: [Int] -> [Int]
shellSort xs = foldl sort xs (gen $ length xs)

-- gap置きにソート(最適化のため)
sort :: [Int] -> Int -> [Int]
sort [] _ = []
-- 本来は、初期値のgapの値も0<=x<gap? 1<=x<=gapで設定する必要あり
sort (x:xs) gap = insert gap x (sort xs gap) where
  insert _ y [] = [y]
  insert g y (z:zs)
    | g /= 0 = z:insert (g-1) y zs
    | z < y = z:insert gap y zs
    | otherwise = y:z:zs

main = do
  print $ sort [3, 1, 2, 5, 4] 0
  print $ shellSort [3, 1, 2, 5, 4]

File: sample/sort/shell-sort.go
package main

import "fmt"

type ShellSort []int

func (x ShellSort) Sort() {
	for gap := len(x) / 2; gap >= 1; gap /= 2 {
		for i := 0; i < len(x); i++ {
			for j := i; j-gap >= 0 && x[j-gap] > x[j]; j -= gap {
				x[j], x[j-gap] = x[j-gap], x[j]
			}
		}
	}
}

func main() {
	var x ShellSort = ShellSort{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 6, 5, 4, 3, 6, 5, 4, 3, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0}
	x.Sort()
	fmt.Println(x)
}
